arangod/Aql/FixedVarExpressionContext.cpp

```diff
@@ -82,36 +89,13 @@ FixedVarExpressionContext::FixedVarExpressionContext(
     AqlFunctionsInternalCache& cache)
     : QueryExpressionContext(trx, context, cache) {}
 
-SingleVarExpressionContext::SingleVarExpressionContext(
-    transaction::Methods& trx, QueryContext& context,
-    AqlFunctionsInternalCache& cache, Variable* var, AqlValue val)
-    : QueryExpressionContext(trx, context, cache),
-      _variable(var),
-      _value(val) {}
-
 SingleVarExpressionContext::SingleVarExpressionContext(
     transaction::Methods& trx, QueryContext& context,
     AqlFunctionsInternalCache& cache)
-    : SingleVarExpressionContext(trx, context, cache, nullptr,
-                                 AqlValue(AqlValueHintNull())) {}
-
-SingleVarExpressionContext::~SingleVarExpressionContext() { _value.destroy(); }
-
-bool SingleVarExpressionContext::isDataFromCollection(Variable const*) const {
-  return false;
-}
-
-AqlValue SingleVarExpressionContext::getVariableValue(Variable const* var, bool,
-                                                      bool&) const {
-  if (var == _variable) {
-    return _value;
-  } else {
-    return AqlValue(AqlValueHintNull());
-  }
-}
+    : QueryExpressionContext(trx, context, cache) {}
 
-void SingleVarExpressionContext::setVariableValue(Variable* variable,
-                                                  AqlValue& value) {
-  _variable = variable;
-  _value = value;
+AqlValue SingleVarExpressionContext::getVariableValue(
+    Variable const* /*variable*/, bool /*doCopy*/,
+    bool& /*mustDestroy*/) const {
+  return AqlValue(AqlValueHintNull());
```

Comment:
why always `Null` for `SingleVarExpressionContext`?

Reply:
As far as I can see, the previous code here was like this:
``` 
AqlValue SingleVarExpressionContext::getVariableValue(Variable const* var, bool,
                                                      bool&) const {
  if (var == _variable) {
    return _value;
  } else {
    return AqlValue(AqlValueHintNull());
  }
}
```
So `_value` was return if `var == _variable`. 
`_variable` however is only initialized in the ctor of `SingleVariableExpressionContext` and in its `setVariableValue()` method.
There were 2 ctors for `SingleVariableExpressionContext` before, and only one of them was used. That ctor initialized `_variable` to nullptr. And there were no callers to `setVariableValue`.
So basically the previous `SingleVariableExpressionContext` was effectively a `NoVariableExpressionContext`.
It is only used from AqlAnalyzers though. I can't tell if this is a bug, or was just missing a cleanup.

Reply:
@jsteemann Ok, that looks like a bug, but let's ask the author himself :-)

Reply:
@alexbakharew any thoughts ^^^?

Reply:
Hi @jsteemann 

Everything is OK with AqlAnalyzers. We had discussion with @Dronplane about SingleVarExpressionContext. And its purpose is to store only one variable against few in FixedVarExpressionContext. 

In process of implementing optimizer rules, we understood that we actually don't want to store value in SingleVarExpressionContext, that's why it is nullptr at the moment. And now it is clear that your suggestion about adding NoVariableExpressionContext is very useful.

Reply:
But I think that we should replace SingleVarExpressionContext with NoVariableExpressionContext or don't modify getVariableValue() method. Because I think that it's very unclear for other developers why we always return nullptr

Reply:
Renamed as suggested.